katcp/core.py

```diff
@@ -1898,3 +1898,48 @@ def maybe_timeout(f):
     maybe_timeout.remaining = _remaining
 
     return maybe_timeout
+
+@tornado.gen.coroutine
+def until_some(*args, **kwargs):
+    """Return a future that resolves when some of the passed futures resolve.
+
+    The futures can be passed as either a sequence of *args* or a dict of
+    *kwargs* (but not both). Some additional keyword arguments are supported,
+    as described below. Once a specified number of underlying futures have
+    resolved, the returned future resolves as well, or a timeout could be
+    raised if specified.
+
+    Parameters
+    ----------
+    done_at_least : None or int
+        Number of futures that need to resolve before this resolves (default all)
+    timeout : None or float
+        Timeout in seconds, or None for no timeout (the default)
+
+    Returns
+    -------
+    This command returns a tornado Future that resolves with a list of
+    (index, value) tuples containing the results of all futures that resolved,
+    with corresponding indices (numbers for *args* futures or keys for *kwargs*
+    futures).
+
+    Raises
+    ------
+    :class:`tornado.gen.TimeoutError`
+        If operation times out before the requisite number of futures resolve
+
+    """
+    done_at_least = kwargs.pop('done_at_least', None)
+    timeout = kwargs.pop('timeout', None)
+    # At this point args and kwargs are either empty or contain futures only
+    if done_at_least is None:
+        done_at_least = len(args) + len(kwargs)
+    wait_iterator = tornado.gen.WaitIterator(*args, **kwargs)
+    maybe_timeout = future_timeout_manager(timeout)
+    results = []
+    while not wait_iterator.done():
+        result = yield maybe_timeout(wait_iterator.next())
+        results.append((wait_iterator.current_index, result))
+        if len(results) >= done_at_least:
+            break
```

Comment:
WaitIterator iterates through the futures as you pass them in [1].
Thus if you have a timeout on the first future (even with maybe_timeout) and all the other were successful it would still result in a future exception. If thats the case you have to catch timeouts and reraise outside the while loop.

[1] Hope to be wrong then it is much simpler for you.

I'll try explaining in code - something like.
```
    your_clear_to_loanch_the_timeout_moab = 0
    while not wait_iterator.done():
         try:
             result = yield maybe_timeout(wait_iterator.next())
         except tornado.gen.TimeoutError:
             your_clear_to_loanch_the_timeout_moab += 1
         results.append((wait_iterator.current_index, result))
         if len(results) >= done_at_least:
             break
    if your_clear_to_loanch_the_timeout_moab:
         raise tornado.gen.TimeoutError("timeout")
```
--------------------
From the code I assume your primary use case is to wait for a handful of wait-until-sensor-value-reaches(X) futures. I like the general nature of this but think for requests it will need to be done slightly differently, taking into account the result.
P.S. I am away for a few days, so @brickZA can decide.